[FLINK-21602] Make ExecutionGraph creation asynchronous in the AdaptiveScheduler

tillrohrmann opened the pull request:
This PR adds functionality to let the `AdaptiveScheduler` create the `ExecutionGraph` asynchronously. This is necessary to avoid blocking I/O operations (retrieving of checkpoints) to happen in the main thread. The change consists essentially of three steps:

1. Introducing a new `CreatingExecutionGraph` state between `WaitingForResources` and `Executing`
2. Splitting the determination of the job parallelism and reserving of slots in two steps
3. Tracking of asynchronous background tasks and waiting on their completion before closing checkpoint services

### New `CreatingExecutionGraph` state

The new `CreatingExecutionGraph` state has the purpose to model that an asynchronous `ExecutionGraph` creation is happening. Since the operation happens asynchronously, it can happen that the available set of slots changes:

* If the `ExecutionGraph` creation succeeds and if we can assign the required slots, then we go to the `Executing` state.
* If the `ExecutionGraph` creation succeeds and we cannot assign the required slots (e.g. some slots were lost), then we go back to the `WaitingForResources` state.
* If the `ExecutionGraph` creation fails, then we go into the `Finished` state and with a `JobStatus.FAILED` result

### Splitting of determining job parallelism and reserving slots

Since we do not directly create the `ExecutionGraph` after we have decided on its parallelism we cannot directly reserve the required slots. In order to support this two-step process, the `SlotAllocator` interface was adjusted to make the `reserveResources` failable. The new API is `Optional<? extends ReservedSlots> tryReserveResources(VertexParallelism vertexParallelism);` which returns `Optional.empty()` if the slots cannot be reserved. This allows the `AdaptiveScheduler` to tell the `CreatingExecutionGraph` state that it could not reserve the required slots which lets the scheduler transition back into the `WaitingForResources` state.

### Tracking of asynchronous tasks

With the introduction of the asynchronous `ExecutionGraph` creation, we also introduced concurrent accesses to the checkpoint services (`CompletedCheckpointStore`, `CheckpointIdCounter`) because they are now accessed by the `ioExecutor` which creates the `ExecutionGraph` and by the main thread when the `Scheduler` reaches the `Finished` state. This is a problem and in order to solve it, the `AdaptiveScheduler` now tracks its background/asynchronous tasks. Moreover, it only allows to run a single task at a time. Before shutting down the checkpoint services, it waits for the completion of these tasks in order to make sure that no concurrent accesses to the `CompletedCheckpointStore` and the `CheckpointIdCounter` happen.

zentol commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/BackgroundTaskTest.java:
```diff
@@ -0,0 +1,148 @@
+import static org.junit.Assert.fail;
+
+/** Tests for the {@link BackgroundTask}. */
+public class BackgroundTaskTest extends TestLogger {
```
We should also add a test that aborting a future that does affect subsequent BackgroundTasks.

tillrohrmann replied:
I don't fully understand which scenario you would like to test. Could you explain it a bit more in detail?

tillrohrmann replied:
Something like `b1 -> b2 -> b3`, blocking on `b1` and then aborting `b2` and checking that after unblocking `b1`, `b3` will be executed?

zentol replied:
I'm interested in the behavior for the case where we have `b1 -> b2`, and `b1` fails.

I initially read the code such that `b2` would still run, but that is not the case because we use `thenApplyAsync`.

This was a bit surprising actually; the javadocs for the BackGroundTask don't make this clear because they don't contain any notion of failed executions. Conceptually I considered this to just be queue that we pull from with some CompletableFuture sugar.
I'm not sure if the behavior is really desired. Shouldn't background tasks that don't have anything to do with each other be independent? If a dependency is required then the result future should be used to make this explicit?

tillrohrmann replied:
I have extended `testAbortSkipsTasksWhichHaveNotBeenStarted` to cover this case. I hope this does what you envisioned.

tillrohrmann replied:
I think the implementation is as you described it. Tasks are independent of each other and if `b1` fails, then `b2` will be executed. The trick is to use the `termination` future which never transports the failure cause but is only completed once the task has completed.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/AdaptiveScheduler.java:
```diff
@@ -713,58 +698,97 @@ public void goToFinished(ArchivedExecutionGraph archivedExecutionGraph) {
+    @Nonnull
+    private ExecutionGraph assignSlotsToExecutionGraph(
+            ExecutionGraph executionGraph,
+            org.apache.flink.runtime.scheduler.adaptive.allocator.ReservedSlots assignedSlots) {
```
Indeed. I will fix it.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/BackgroundTask.java:
```diff
@@ -0,0 +1,128 @@
+
+    private final CompletableFuture<T> resultFuture;
+
+    private volatile boolean isValid = true;
```
fair enough. I will change it.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/CreatingExecutionGraph.java:
```diff
@@ -0,0 +1,228 @@
+        }
+
+        static AssignmentResult success(ExecutionGraph executionGraph) {
+            return new AssignmentResult(executionGraph);
```
I will add it.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/allocator/SlotAllocator.java:
```diff
@@ -45,27 +42,26 @@
+     *     Optional#empty()}
      */
-    Map<ExecutionVertexID, LogicalSlot> reserveResources(T vertexParallelism);
+    Optional<? extends ReservedSlots> tryReserveResources(VertexParallelism vertexParallelism);
```
The idea is to have a way to tell the caller that the `SlotAllocator` could not reserve the required slots for the given `VertexParallelism`. The alternative would be to return an empty `ReservedSlots` instance but this would be a bit less expressive but also possible.

zentol replied:
my bad, I meant why it is returning `Optional<? extends ReservedSlots>` instead of `Optional<ReservedSlots>`. The class can't be extended due to the private constructor.

tillrohrmann replied:
Fair point. I will remove the `? extends`.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/allocator/SlotSharingSlotAllocator.java:
```diff
@@ -148,7 +149,17 @@ public ResourceCounter calculateRequiredSlots(
-            VertexParallelismWithSlotSharing vertexParallelismWithSlotSharing) {
+            VertexParallelism vertexParallelism) {
+        Preconditions.checkArgument(
+                vertexParallelism instanceof VertexParallelismWithSlotSharing,
```
True, this should also work. Given that this is an internal API and that changing it would require some extra work, I would keep it as is and maybe change it in the future if we want.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/AdaptiveSchedulerTest.java:
```diff
@@ -722,25 +727,50 @@ public void testRestoringModifiedJobFromSavepointFails() throws Exception {
+
+        try {
+            final ComponentMainThreadExecutor singleMainThreadExecutor =
+                    ComponentMainThreadExecutorServiceAdapter.forSingleThreadExecutor(
```
I think the important bit is to use the manually triggered main thread executor as little as possible. This makes things just harder to change in the future. If the tests are build so that they can work together with a normal `Executor`, then it is much better. In fact, I start to believe that tests which require the manually triggered main thread executor are smelling a bit. This can be an indicator that the class under test is not really well testable.

zentol replied:
I agree that we should reduce usages of the manual executor, but that wasn't the direction I intended to go.
I was more thinking along the lines of `ComponentMainThreadExecutorServiceAdapter.forMainThread()`

tillrohrmann replied:
Also this executor is not ideal because it produces thread interleavings which cannot happen with a normal main thread executor.

tillrohrmann commented:
Thanks for the review @zentol. I think I have addressed most of your comments and answered those which were more of a discussion. Please take another look.

zentol reviewed (approved):
+1
